### DISCUSSION TITLE
Getting a timeout within an action on `run ssh`

### DISCUSSION THREAD

**Discussion by PeteAUK** [community/community#51501] [Actions] [answered]
### Select Topic Area

Question

### Body

Not sure if this is something I'm doing wrong or a bug if I'm honest.  What I'm trying to do is use an Action to SSH into my server and run some commands.  This is with GitHub's standard runners and my repository is private.

My YAML file is as follows:
```name: 🚀 SSH Deploy
on:
  push:
    branches:
      - main
  workflow_dispatch:
  
jobs:
  run_pull:
    name: run pull
    runs-on: ubuntu-latest
    
    steps:
    - name: 🖊 Write SSH keys
      run: |
        install -m 600 -D /dev/null ~/.ssh/id_rsa
        echo "${{ secrets.SSH_PRIVATE_KEY }}" > ~/.ssh/id_rsa
        ssh-keyscan -H ${{ secrets.SSH_HOST }} > ~/.ssh/known_hosts
  
    - name: 🔗 Connect and Pull
      run: ssh ${{ secrets.SSH_USER }}@${{ secrets.SSH_HOST }} "cd ${{ env.WORK_DIR }} && git checkout ${{ env.MAIN_BRANCH }} && git pull && composer update && exit"

    - name: 🧹 Cleanup
      run: rm -rf ~/.ssh
```

Each time I run this I receive `ssh: connect to host *** port 22: Connection timed out`, which occurs around the 2 minute mark of the action firing off.

I have connected to my server both from PuTTY and the command line without any issues (which I believe shows the keys are installed on my server OK).  I have also re-entered my secrets on multiple occasions and verified they're under Actions and not one of the other stores.

I've even tried `ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no` to see if that was what was causing the problem and got a timeout with this.  I've not tried changing any of my commands because the ssh connection isn't establishing at all as far as I can tell.

Is there something special I need to do - it feels like the runner's access to my server is blocked which I can't quite explain?

**Comment by airtower-luna** [comment, marked as answer]
Timeout (as opposed to a rejected connection) sounds very much like a firewall is dropping the packets. I don't know the network environment of your server, but that's the first thing I'd check. Mind that the firewall doesn't have to be the one of the server itself, could also be a router in between.

Though it'd be odd for it to work with `ssh-keyscan` and then not with `ssh`, at exactly what point does the error occur? Adding `-v` might be useful either way.

Finally, using `ssh-keyscan` like that is unsafe. If someone is able to do an MITM attack on the runner, it'll likely see the same (attacker) key as the actual SSH connection. Better store the host public keys in a secret (or simply a file in your repository) and use that.

**Reply by PeteAUK** [reply]
Thanks for the reply :)  The error occurs at the connect and pull step, I've just added the verbose flag and got this:

```debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19 include /etc/ssh/ssh_config.d/*.conf matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug1: Connecting to *** [<myserverip>] port 22.
debug1: connect to address <myserverip> port 22: Connection timed out
ssh: connect to host *** port 22: Connection timed out
Error: Process completed with exit code 255.
```

I'm not sure why it's connecting via IP rather than domain name, but I've tested this on my local computer and it works OK.  That's sounding like the host I'm using may be blocking access to the runner 🤔

That's interesting on the `ssh-keyscan`, I largely picked most of this up from searching online and the pitfalls of snapshotting it was that you needed to maintain that whereas the keyscan sorted itself out on the fly.

**Reply by airtower-luna** [reply]
> I'm not sure why it's connecting via IP rather than domain name

Well, it needs to resolve the name to an IP address to actually connect. :wink: 

> That's interesting on the ssh-keyscan, I largely picked most of this up from searching online and the pitfalls of snapshotting it was that you needed to maintain that whereas the keyscan sorted itself out on the fly.

If you use `ssh-keyscan` on a typical desktop system where your home directory is persistent that's effectively TOFU (trust on first use). You *assume* that no-one is manipulating the first connection you make to get the key, and if the key then changes you know that something is wrong. Not ideal security-wise, but often sufficient (depending on your threat model, it's definitely not sufficient if the NSA or someone similar is after you personally).

But using `ssh-keyscan` during every workflow run effectively disables the host key check. *If* someone is messing with the connection to your server during the job, they'll probably manipulate the two connections (keyscan and following `ssh` command) alike.

**Reply by PeteAUK** [reply]
> > I'm not sure why it's connecting via IP rather than domain name
> 
> Well, it needs to resolve the name to an IP address to actually connect. 😉

I know that 😉  It was more that it looked like the ssh was being attempted via IP address rather than host name.  Could just be the way it's being displayed in truth.

I'm asking my hosting company if they're aware of anything that could cause this.  I have a horrible feeling there's something in play that isn't making itself obvious as everything should work out the box.

> > That's interesting on the ssh-keyscan, I largely picked most of this up from searching online and the pitfalls of snapshotting it was that you needed to maintain that whereas the keyscan sorted itself out on the fly.
> 
> If you use `ssh-keyscan` on a typical desktop system where your home directory is persistent that's effectively TOFU (trust on first use). You _assume_ that no-one is manipulating the first connection you make to get the key, and if the key then changes you know that something is wrong. Not ideal security-wise, but often sufficient (depending on your threat model, it's definitely not sufficient if the NSA or someone similar is after you personally).
> 
> But using `ssh-keyscan` during every workflow run effectively disables the host key check. _If_ someone is messing with the connection to your server during the job, they'll probably manipulate the two connections (keyscan and following `ssh` command) alike.

In truth it sounds like the likelihood of somebody doing this is slim, however if I can sort out the issue with getting the runner to ssh onto my server I'll change this over to a secure secret.

**Reply by PeteAUK** [reply]
Ahha!  I've sorted it.  It would appear that `ssh-keyscan` on the runner generated a different output than on a local linux machine...  Passing it in as a secret sorted that out, so amusingly your security suggestion was actually the fix 😂  What is weird is telling it not check the hosts file didn't seem to work when I tried it.

Couple of other tweaks in that I was using `env.` and not `vars.`, but all working neatly now.  Just need to work out how to write a commit version or similar so that I can clearly see what code the site is running on.

**Reply by airtower-luna** [reply]
Glad to hear it works now! :smile: 

> Just need to work out how to write a commit version or similar so that I can clearly see what code the site is running on.

Take a look at `git rev-parse HEAD`. :wink:

**Reply by pstepanek954** [reply]
Peto di do pici ty kundo cau misa

**Reply by PeteAUK** [reply]
> Glad to hear it works now! 😄
> 
> > Just need to work out how to write a commit version or similar so that I can clearly see what code the site is running on.
> 
> Take a look at `git rev-parse HEAD`. 😉

Thanks!  Was just realising I was overthinking it and you came up with the answer 😁

**Comment by pstepanek954** [comment]
Neboj uz odlitam tak uz mas klid uz tr nechci videt  sem stebou skoncila
